sympy/tensor/functions.py

```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
             arg.__getitem__(tuple(next(index) for i in shp))
             for arg, shp in zip(self.args, self._get_args_shapes())
         )
+
+
+def shape(expr):
+    """
+    Return the shape of the tensor.
+
+    This function dispatches over the kind and class of *expr* to return
+    its shape.
+
+    Examples
+    ========
+
+    >>> from sympy import shape, Array, Matrix, Integral
+    >>> from sympy.abc import x
+    >>> A = Array([1, 2])
+    >>> shape(A)
+    (2,)
+    >>> shape(Integral(A, x))
+    (2,)
+
+    Scalars return empty tuple since they are rank-0 tensor.
+
+    >>> shape(x)
+    ()
+
+    Shape of the matrix has two elements.
+
+    >>> M = Matrix([1, 2])
+    >>> shape(M)
+    (2, 1)
+    >>> shape(Integral(M, x))
+    (2, 1)
+
+    """
+    return accept_shape(expr.kind, expr)
+
+@singledispatch
+def accept_shape(kind, expr):
+    raise NotImplementedError
+
+@accept_shape.register(_NumberKind)
+def _(kind, expr):
+    return ()
```

Comment:
Is it useful to have an empty tuple as shape for a scalar?

I would expect it to be more useful to get an error since this will usually be a bug.